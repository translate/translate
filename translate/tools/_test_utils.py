import sys

import pytest

requires_py38_mark = pytest.mark.skipif(
    sys.version_info < (3, 8), reason="Requires python 3.8"
)

test_po_files = [
    "tests/cli/data/test_pocount_po_csv/one.po",
    "tests/cli/data/test_pocount_po_file/one.po",
    "tests/cli/data/test_pocount_po_fuzzy/one.po",
]
