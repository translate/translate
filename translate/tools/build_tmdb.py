#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2008 Zuza Software Foundation
#
# This file is part of the Translate Toolkit.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.

"""Import units from translations files into tmdb."""

import logging
import os
from argparse import ArgumentParser

from translate.storage import factory, tmdb


logger = logging.getLogger(__name__)


class Builder:

    def __init__(self, tmdbfile, source_lang, target_lang, filenames):
        self.tmdb = tmdb.TMDB(tmdbfile)
        self.source_lang = source_lang
        self.target_lang = target_lang

        for filename in filenames:
            if not os.path.exists(filename):
                logger.error("cannot process %s: does not exist", filename)
                continue
            elif os.path.isdir(filename):
                self.handledir(filename)
            else:
                self.handlefile(filename)
        self.tmdb.connection.commit()

    def handlefile(self, filename):
        try:
            store = factory.getobject(filename)
        except Exception as e:
            logger.error(str(e))
            return
        # do something useful with the store and db
        try:
            self.tmdb.add_store(store, self.source_lang, self.target_lang, commit=False)
        except Exception as e:
            print(e)
        print("File added:", filename)

    def handlefiles(self, dirname, filenames):
        for filename in filenames:
            pathname = os.path.join(dirname, filename)
            if os.path.isdir(pathname):
                self.handledir(pathname)
            else:
                self.handlefile(pathname)

    def handledir(self, dirname):
        path, name = os.path.split(dirname)
        if name in ["CVS", ".svn", "_darcs", ".git", ".hg", ".bzr"]:
            return
        entries = os.listdir(dirname)
        self.handlefiles(dirname, entries)


def main():
    parser = ArgumentParser()
    parser.add_argument(
        "-d", "--tmdb", dest="tmdb_file", default="tm.db",
        help="translation memory database file (default: tm.db)")
    parser.add_argument(
        "-s", "--import-source-lang", dest="source_lang", default="en",
        help="source language of translation files (default: en)")
    parser.add_argument(
        "-t", "--import-target-lang", dest="target_lang",
        help="target language of translation files", required=True)
    parser.add_argument(
        "files", metavar="input files", nargs="+"
    )
    args = parser.parse_args()

    logging.basicConfig(format="%(name)s: %(levelname)s: %(message)s")

    Builder(args.tmdb_file, args.source_lang, args.target_lang, args.files)

if __name__ == '__main__':
    main()
