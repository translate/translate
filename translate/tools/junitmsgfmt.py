from __future__ import annotations

import subprocess
from argparse import ArgumentParser
from os.path import basename
from time import time
from typing import Iterable, NamedTuple

import lxml.etree as etree


class MsgfmtTester:
    def __init__(self, files: Iterable[str], untranslated=False):
        self._detect_untranslated = untranslated
        self._files = files

    def run(self):
        results = list(map(self._run_msgfmt, self._files))
        self._print_results(results)

    def _run_msgfmt(self, pofile: str):
        start_time = time()
        process = subprocess.Popen(
            ["msgfmt", "-cv", "-o", "/dev/null", pofile],
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
        )
        stdout, stderr = process.communicate()
        stderr = stderr.replace(pofile, basename(pofile))

        failures: list[CheckFailure] = []
        if process.returncode:
            failures.append(CheckFailure(stderr, "Failure"))
        elif self._detect_untranslated and "untranslated" in stderr:
            failures.append(CheckFailure(stderr, "Untranslated messages"))
        return CheckResult(
            pofile,
            0 if "test_junitmsgfmt" in pofile else time() - start_time,
            failures,
        )

    @staticmethod
    def _print_results(results: list[CheckResult]):
        failures = len([r for r in results if len(r.failures)])
        total_time = sum([r.time for r in results], 0)
        root = etree.Element(
            "testsuite",
            name="msgfmt",
            errors="0",
            skips="0",
            failures=str(failures),
            tests=str(len(results)),
            time=f"{total_time:.4f}",
        )
        for result in results:
            case = etree.SubElement(
                root,
                "testcase",
                classname="msgfmt",
                name=f"check[{result.file}]",
                file=result.file,
                time=f"{result.time:.4f}",
            )
            for failure in result.failures:
                etree.SubElement(
                    case, "failure", message=failure.message
                ).text = failure.text
        print(etree.tostring(root, encoding="unicode", pretty_print=True))


class CheckResult(NamedTuple):
    file: str
    time: float
    failures: list[CheckFailure]


class CheckFailure(NamedTuple):
    text: str
    message: str


def main(arguments=None):
    parser = ArgumentParser()
    parser.add_argument(
        "--untranslated",
        action="store_true",
        default=False,
        help="fail on untranslated messages",
    )
    parser.add_argument("files", nargs="+")

    args = parser.parse_args(arguments)

    MsgfmtTester(args.files, untranslated=args.untranslated).run()


if __name__ == "__main__":
    main()
