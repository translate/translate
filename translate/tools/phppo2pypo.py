#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2009 Mozilla Corporation, Zuza Software Foundation
#
# This file is part of translate.
#
# translate is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# translate is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.

"""Convert PHP format .po files to Python format .po files.
"""

import re

from translate.storage import po
from translate.misc.multistring import multistring


class phppo2pypo:

    def convertstore(self, inputstore):
        """Converts a given .po file (PHP Format) to a Python format .po file, the difference being
            how variable substitutions work.  PHP uses a %1$s format, and Python uses
            a {0} format (zero indexed).  This method will convert, e.g.:
                I have %2$s apples and %1$s oranges
                    to
                I have {1} apples and {0} oranges
            This method ignores strings with %s as both languages will recognize that.
        """
        thetargetfile = po.pofile(inputfile="")

        for unit in inputstore.units:
            newunit = self.convertunit(unit)
            thetargetfile.addunit(newunit)
        return thetargetfile

    def convertunit(self, unit):
        developer_notes = unit.getnotes(origin="developer")
        translator_notes = unit.getnotes(origin="translator")
        unit.removenotes()
        unit.addnote(self.convertstrings(developer_notes))
        unit.addnote(self.convertstrings(translator_notes))
        unit.source = self.convertstrings(unit.source)
        unit.target = self.convertstrings(unit.target)
        return unit

    def convertstring(self, input):
        return re.sub('%(\d)\$s', lambda x: "{%d}" % (int(x.group(1)) - 1), input)

    def convertstrings(self, input):
        if isinstance(input, multistring):
            strings = input.strings
        elif isinstance(input, list):
            strings = input
        else:
            return self.convertstring(input)
        for index, string in enumerate(strings):
            strings[index] = re.sub('%(\d)\$s', lambda x: "{%d}" % (int(x.group(1)) - 1), string)
        return multistring(strings)


def convertphp2py(inputfile, outputfile, template=None):
    """Converts from PHP .po format to Python .po format

    :param inputfile: file handle of the source
    :param outputfile: file handle to write to
    :param template: unused
    """
    convertor = phppo2pypo()
    inputstore = po.pofile(inputfile)
    outputstore = convertor.convertstore(inputstore)
    if outputstore.isempty():
        return False
    outputfile.write(str(outputstore))
    return True


def main(argv=None):
    """Converts PHP .po files to Python .po files."""
    from translate.convert import convert

    formats = {"po": ("po", convertphp2py)}
    parser = convert.ConvertOptionParser(formats, description=__doc__)
    parser.run(argv)


if __name__ == '__main__':
    main()
