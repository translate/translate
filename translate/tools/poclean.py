#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2007 Zuza Software Foundation
#
# This file is part of translate.
#
# translate is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# translate is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.

"""Produces a clean file from an unclean file (Trados/Wordfast) by stripping
out the tw4win indicators.

This does not convert an RTF file to PO/XLIFF, but produces the target file
with only the target text in from a text version of the RTF.
"""

import re

from translate.misc.multistring import multistring
from translate.storage import factory


tw4winre = re.compile(r"\{0>.*?<\}\d{1,3}\{>(.*?)<0\}", re.M | re.S)


def cleanunit(unit):
    """cleans the targets in the given unit"""
    if isinstance(unit.target, multistring):
        strings = unit.target.strings
    else:
        strings = [unit.target]
    for index, string in enumerate(strings):
        string = string.replace("\par", "")
        strings[index] = tw4winre.sub(r"\1", string)
    if len(strings) == 1:
        unit.target = strings[0]
    else:
        unit.target = strings


def cleanfile(thefile):
    """cleans the given file"""
    for unit in thefile.units:
        cleanunit(unit)
    return thefile


def runclean(inputfile, outputfile, templatefile):
    """reads in inputfile, cleans, writes to outputfile"""
    fromfile = factory.getobject(inputfile)

    cleanfile(fromfile)
#    if fromfile.isempty():
#        return False
    outputfile.write(str(fromfile))
    return True


def main():
    from translate.convert import convert
    formats = {"po": ("po", runclean), "xlf": ("xlf", runclean), None: ("po", runclean)}
    parser = convert.ConvertOptionParser(formats, usetemplates=False, description=__doc__)
    parser.run()


if __name__ == '__main__':
    main()
