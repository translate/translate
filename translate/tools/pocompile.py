#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright 2005, 2006,2010 Zuza Software Foundation
#
# This file is part of the translate-toolkit
#
# translate is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# translate is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, see <http://www.gnu.org/licenses/>.

"""Compile XLIFF and Gettext PO localization files into Gettext MO (Machine Object) files.

See: http://docs.translatehouse.org/projects/translate-toolkit/en/latest/commands/pocompile.html
for examples and usage instructions.
"""

from translate.storage import factory
from translate.storage import mo
from translate.misc.multistring import multistring


def _do_msgidcomment(string):
    return u"_: %s\n" % string


class POCompile:

    def convertstore(self, inputfile, includefuzzy=False):
        outputfile = mo.mofile()
        for unit in inputfile.units:
            if unit.istranslated() or (unit.isfuzzy() and includefuzzy and unit.target) or unit.isheader():
                mounit = mo.mounit()
                if unit.isheader():
                    mounit.source = ""
                else:
                    mounit.source = unit.source
                    context = unit.getcontext()
                    if unit.msgidcomment:
                        if mounit.hasplural():
                            mounit.source = multistring(_do_msgidcomment(unit.msgidcomment) + mounit.source, *mounit.source.strings[1:])
                        else:
                            mounit.source = _do_msgidcomment(unit.msgidcomment) + mounit.source
                    elif context:
                        mounit.msgctxt = [context]
                mounit.target = unit.target
                outputfile.addunit(mounit)
        return str(outputfile)


def convertmo(inputfile, outputfile, templatefile, includefuzzy=False):
    """reads in a base class derived inputfile, converts using pocompile, writes to outputfile"""
    # note that templatefile is not used, but it is required by the converter...
    inputstore = factory.getobject(inputfile)
    if inputstore.isempty():
        return 0
    convertor = POCompile()
    outputmo = convertor.convertstore(inputstore, includefuzzy)
    # We have to make sure that we write the files in binary mode, therefore we
    # reopen the file accordingly
    outputfile.close()
    outputfile = open(outputfile.name, 'wb')
    outputfile.write(outputmo)
    return 1


def main():
    from translate.convert import convert
    formats = {"po": ("mo", convertmo), "xlf": ("mo", convertmo)}
    parser = convert.ConvertOptionParser(formats, usepots=False, description=__doc__)
    parser.add_fuzzy_option()
    parser.run()


if __name__ == '__main__':
    main()
