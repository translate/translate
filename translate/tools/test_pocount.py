from io import BytesIO

from pytest import mark

from translate.storage import po
from translate.tools import pocount


class TestCount:
    def count(self, source, expectedsource, target=None, expectedtarget=None):
        """simple helper to check the respective word counts"""
        poelement = po.pounit(source)
        if target is not None:
            poelement.target = target
        wordssource, wordstarget = pocount.wordsinunit(poelement)
        print(
            'Source (expected=%d; actual=%d): "%s"'
            % (expectedsource, wordssource, source)
        )
        assert wordssource == expectedsource
        if target is not None:
            print(
                'Target (expected=%d; actual=%d): "%s"'
                % (expectedtarget, wordstarget, target)
            )
            assert wordstarget == expectedtarget

    def test_simple_count_zero(self):
        """no content"""
        self.count("", 0)

    def test_simple_count_one(self):
        """simplest one word count"""
        self.count("One", 1)

    def test_simple_count_two(self):
        """simplest one word count"""
        self.count("One two", 2)

    def test_punctuation_divides_words(self):
        """test that we break words when there is punctuation"""
        self.count("One. Two", 2)
        self.count("One.Two", 2)

    def test_xml_tags(self):
        """test that we do not count XML tags as words"""
        # <br> is a word break
        self.count("A word<br>Another word", 4)
        self.count("A word<br/>Another word", 4)
        self.count("A word<br />Another word", 4)
        # \n is a word break
        self.count("<p>A word</p>\n<p>Another word</p>", 4)
        # Not really an XML tag
        self.count("<no label>", 2)

    def test_newlines(self):
        """test to see that newlines divide words"""
        # newlines break words
        self.count("A word.\nAnother word", 4)
        self.count(r"A word.\\n\nAnother word", 4)

    def test_variables_are_words(self):
        """test that we count variables as words"""
        self.count("%PROGRAMNAME %PROGRAM% %s $file $1", 5)

    def test_plurals(self):
        """test that we can handle plural PO elements"""
        # #: gdk-pixbuf/gdk-pixdata.c:430
        # #, c-format
        # msgid "failed to allocate image buffer of %u byte"
        # msgid_plural "failed to allocate image buffer of %u bytes"
        # msgstr[0] "e paletšwe go hwetša sešireletši sa seswantšho sa paete ya %u"
        # msgstr[1] "e paletšwe go hwetša sešireletši sa seswantšho sa dipaete tša %u"

    @mark.xfail(reason="Support commented out pending removal")
    def test_plurals_kde(self):
        """test that we correcly count old style KDE plurals"""
        self.count("_n: Singular\\n\nPlural", 2, "Een\\n\ntwee\\n\ndrie", 3)

    def test_msgid_blank(self):
        """counts a message id"""
        self.count("   ", 0)

    # Counting strings
    #  We need to check how we count strings also and if we call it translated or untranslated
    # ie an all spaces msgid should be translated if there are spaces in the msgstr

    # Make sure we don't count obsolete messages

    # Do we correctly identify a translated yet blank message?

    # Need to test that we can differentiate between fuzzy, translated and untranslated


class TestPOCount:
    """
    This only tests the old (memory-based) pocount method.
    """

    inputdata = br"""
msgid "translated unit"
msgstr "translated unit"

#, fuzzy
msgid "fuzzy unit"
msgstr "fuzzy unit"

# untranslated
msgid "untranslated unit"
msgstr ""

# obsolete
#~ msgid "obsolete translated unit"
#~ msgstr "obsolete translated unit"

#, fuzzy
#~ msgid "obsolete fuzzy unit"
#~ msgstr "obsolete fuzzy unit"

# untranslated
#~ msgid "obsolete untranslated unit"
#~ msgstr ""
"""

    def test_translated(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["translated"] == 1

    def test_fuzzy(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["fuzzy"] == 1

    def test_untranslated(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["untranslated"] == 1

    def test_total(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["total"] == 3

    def test_translatedsourcewords(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["translatedsourcewords"] == 2

    def test_fuzzysourcewords(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["fuzzysourcewords"] == 2

    def test_untranslatedsourcewords(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["untranslatedsourcewords"] == 2

    def test_totalsourcewords(self):
        pofile = BytesIO(self.inputdata)
        stats = pocount.calcstats(pofile)
        assert stats["totalsourcewords"] == 6
