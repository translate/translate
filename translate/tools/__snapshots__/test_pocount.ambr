# serializer version: 1
# name: test_output[no-color=False-incomplete=False-csv]
  '''
  Filename,Translated Messages,Translated Source Words,Translated Target Words,Fuzzy Messages,Fuzzy Source Words,Untranslated Messages,Untranslated Source Words,Total Message,Total Source Words,Review Messages,Review Source Words
  tests/cli/data/test_pocount_po_csv/one.po,1,3,3,0,0,0,0,1,3,0,0
  tests/cli/data/test_pocount_po_file/one.po,1,3,3,0,0,0,0,1,3,0,0
  tests/cli/data/test_pocount_po_fuzzy/one.po,1,2,3,1,2,0,0,2,4,0,0
  
  '''
# ---
# name: test_output[no-color=False-incomplete=False-full]
  '''
  Processing file : [95mtests/cli/data/test_pocount_po_csv/one.po[0m
  Type               Strings      Words (source)    Words (translation)
  [92mTranslated:       1 (100%)          3 (100%)               3[0m
  [93mFuzzy:            0 (  0%)          0 (  0%)             n/a[0m
  [91mUntranslated:     0 (  0%)          0 (  0%)             n/a[0m
  Total:            1                 3                      3
  
  Unreviewed:       1 (100%)          3 (100%)               3
  
  Processing file : [95mtests/cli/data/test_pocount_po_file/one.po[0m
  Type               Strings      Words (source)    Words (translation)
  [92mTranslated:       1 (100%)          3 (100%)               3[0m
  [93mFuzzy:            0 (  0%)          0 (  0%)             n/a[0m
  [91mUntranslated:     0 (  0%)          0 (  0%)             n/a[0m
  Total:            1                 3                      3
  
  Unreviewed:       1 (100%)          3 (100%)               3
  
  Processing file : [95mtests/cli/data/test_pocount_po_fuzzy/one.po[0m
  Type               Strings      Words (source)    Words (translation)
  [92mTranslated:       1 ( 50%)          2 ( 50%)               3[0m
  [93mFuzzy:            1 ( 50%)          2 ( 50%)             n/a[0m
  [91mUntranslated:     0 (  0%)          0 (  0%)             n/a[0m
  Total:            2                 4                      3
  
  Unreviewed:       1 ( 50%)          2 ( 50%)               3
  Needs-Work:       1 ( 50%)          2 ( 50%)               0
  
  Processing file : [95mTOTAL:[0m
  Type               Strings      Words (source)    Words (translation)
  [92mTranslated:       3 ( 75%)          8 ( 80%)               9[0m
  [93mFuzzy:            1 ( 25%)          2 ( 20%)             n/a[0m
  [91mUntranslated:     0 (  0%)          0 (  0%)             n/a[0m
  Total:            4                10                      9
  
  File count:       3
  
  
  '''
# ---
# name: test_output[no-color=False-incomplete=False-short-strings]
  '''
  [95mtests/cli/data/test_pocount_po_csv/one.po[0m   strings: total: 1	| [92m1[0mt	[93m0[0mf	[91m0[0mu	| [92m100.0%[0mt	[93m0.0%[0mf	[91m0.0%[0mu
  [95mtests/cli/data/test_pocount_po_file/one.po[0m  strings: total: 1	| [92m1[0mt	[93m0[0mf	[91m0[0mu	| [92m100.0%[0mt	[93m0.0%[0mf	[91m0.0%[0mu
  [95mtests/cli/data/test_pocount_po_fuzzy/one.po[0m strings: total: 2	| [92m1[0mt	[93m1[0mf	[91m0[0mu	| [92m50.0%[0mt	[93m50.0%[0mf	[91m0.0%[0mu
  
  '''
# ---
# name: test_output[no-color=False-incomplete=False-short-words]
  '''
  [95mtests/cli/data/test_pocount_po_csv/one.po[0m   source words: total: 3	| [92m3[0mt	[93m0[0mf	[91m0[0mu	| [92m100.0%[0mt	[93m0.0%[0mf	[91m0.0%[0mu
  [95mtests/cli/data/test_pocount_po_file/one.po[0m  source words: total: 3	| [92m3[0mt	[93m0[0mf	[91m0[0mu	| [92m100.0%[0mt	[93m0.0%[0mf	[91m0.0%[0mu
  [95mtests/cli/data/test_pocount_po_fuzzy/one.po[0m source words: total: 4	| [92m2[0mt	[93m2[0mf	[91m0[0mu	| [92m50.0%[0mt	[93m50.0%[0mf	[91m0.0%[0mu
  
  '''
# ---
# name: test_output[no-color=False-incomplete=True-csv]
  '''
  Filename,Translated Messages,Translated Source Words,Translated Target Words,Fuzzy Messages,Fuzzy Source Words,Untranslated Messages,Untranslated Source Words,Total Message,Total Source Words,Review Messages,Review Source Words
  tests/cli/data/test_pocount_po_fuzzy/one.po,1,2,3,1,2,0,0,2,4,0,0
  
  '''
# ---
# name: test_output[no-color=False-incomplete=True-full]
  '''
  Processing file : [95mtests/cli/data/test_pocount_po_fuzzy/one.po[0m
  Type               Strings      Words (source)    Words (translation)
  [92mTranslated:       1 ( 50%)          2 ( 50%)               3[0m
  [93mFuzzy:            1 ( 50%)          2 ( 50%)             n/a[0m
  [91mUntranslated:     0 (  0%)          0 (  0%)             n/a[0m
  Total:            2                 4                      3
  
  Unreviewed:       1 ( 50%)          2 ( 50%)               3
  Needs-Work:       1 ( 50%)          2 ( 50%)               0
  
  Processing file : [95mTOTAL (incomplete only):[0m
  Type               Strings      Words (source)    Words (translation)
  [92mTranslated:       3 ( 75%)          8 ( 80%)               9[0m
  [93mFuzzy:            1 ( 25%)          2 ( 20%)             n/a[0m
  [91mUntranslated:     0 (  0%)          0 (  0%)             n/a[0m
  Total:            4                10                      9
  
  File count (incomplete):       1
  File count:       3
  
  
  '''
# ---
# name: test_output[no-color=False-incomplete=True-short-strings]
  '''
  [95mtests/cli/data/test_pocount_po_fuzzy/one.po[0m strings: total: 2	| [92m1[0mt	[93m1[0mf	[91m0[0mu	| [92m50.0%[0mt	[93m50.0%[0mf	[91m0.0%[0mu
  
  '''
# ---
# name: test_output[no-color=False-incomplete=True-short-words]
  '''
  [95mtests/cli/data/test_pocount_po_fuzzy/one.po[0m source words: total: 4	| [92m2[0mt	[93m2[0mf	[91m0[0mu	| [92m50.0%[0mt	[93m50.0%[0mf	[91m0.0%[0mu
  
  '''
# ---
# name: test_output[no-color=True-incomplete=False-csv]
  '''
  Filename,Translated Messages,Translated Source Words,Translated Target Words,Fuzzy Messages,Fuzzy Source Words,Untranslated Messages,Untranslated Source Words,Total Message,Total Source Words,Review Messages,Review Source Words
  tests/cli/data/test_pocount_po_csv/one.po,1,3,3,0,0,0,0,1,3,0,0
  tests/cli/data/test_pocount_po_file/one.po,1,3,3,0,0,0,0,1,3,0,0
  tests/cli/data/test_pocount_po_fuzzy/one.po,1,2,3,1,2,0,0,2,4,0,0
  
  '''
# ---
# name: test_output[no-color=True-incomplete=False-full]
  '''
  Processing file : tests/cli/data/test_pocount_po_csv/one.po
  Type               Strings      Words (source)    Words (translation)
  Translated:       1 (100%)          3 (100%)               3
  Fuzzy:            0 (  0%)          0 (  0%)             n/a
  Untranslated:     0 (  0%)          0 (  0%)             n/a
  Total:            1                 3                      3
  
  Unreviewed:       1 (100%)          3 (100%)               3
  
  Processing file : tests/cli/data/test_pocount_po_file/one.po
  Type               Strings      Words (source)    Words (translation)
  Translated:       1 (100%)          3 (100%)               3
  Fuzzy:            0 (  0%)          0 (  0%)             n/a
  Untranslated:     0 (  0%)          0 (  0%)             n/a
  Total:            1                 3                      3
  
  Unreviewed:       1 (100%)          3 (100%)               3
  
  Processing file : tests/cli/data/test_pocount_po_fuzzy/one.po
  Type               Strings      Words (source)    Words (translation)
  Translated:       1 ( 50%)          2 ( 50%)               3
  Fuzzy:            1 ( 50%)          2 ( 50%)             n/a
  Untranslated:     0 (  0%)          0 (  0%)             n/a
  Total:            2                 4                      3
  
  Unreviewed:       1 ( 50%)          2 ( 50%)               3
  Needs-Work:       1 ( 50%)          2 ( 50%)               0
  
  Processing file : TOTAL:
  Type               Strings      Words (source)    Words (translation)
  Translated:       3 ( 75%)          8 ( 80%)               9
  Fuzzy:            1 ( 25%)          2 ( 20%)             n/a
  Untranslated:     0 (  0%)          0 (  0%)             n/a
  Total:            4                10                      9
  
  File count:       3
  
  
  '''
# ---
# name: test_output[no-color=True-incomplete=False-short-strings]
  '''
  tests/cli/data/test_pocount_po_csv/one.po   strings: total: 1	| 1t	0f	0u	| 100.0%t	0.0%f	0.0%u
  tests/cli/data/test_pocount_po_file/one.po  strings: total: 1	| 1t	0f	0u	| 100.0%t	0.0%f	0.0%u
  tests/cli/data/test_pocount_po_fuzzy/one.po strings: total: 2	| 1t	1f	0u	| 50.0%t	50.0%f	0.0%u
  
  '''
# ---
# name: test_output[no-color=True-incomplete=False-short-words]
  '''
  tests/cli/data/test_pocount_po_csv/one.po   source words: total: 3	| 3t	0f	0u	| 100.0%t	0.0%f	0.0%u
  tests/cli/data/test_pocount_po_file/one.po  source words: total: 3	| 3t	0f	0u	| 100.0%t	0.0%f	0.0%u
  tests/cli/data/test_pocount_po_fuzzy/one.po source words: total: 4	| 2t	2f	0u	| 50.0%t	50.0%f	0.0%u
  
  '''
# ---
# name: test_output[no-color=True-incomplete=True-csv]
  '''
  Filename,Translated Messages,Translated Source Words,Translated Target Words,Fuzzy Messages,Fuzzy Source Words,Untranslated Messages,Untranslated Source Words,Total Message,Total Source Words,Review Messages,Review Source Words
  tests/cli/data/test_pocount_po_fuzzy/one.po,1,2,3,1,2,0,0,2,4,0,0
  
  '''
# ---
# name: test_output[no-color=True-incomplete=True-full]
  '''
  Processing file : tests/cli/data/test_pocount_po_fuzzy/one.po
  Type               Strings      Words (source)    Words (translation)
  Translated:       1 ( 50%)          2 ( 50%)               3
  Fuzzy:            1 ( 50%)          2 ( 50%)             n/a
  Untranslated:     0 (  0%)          0 (  0%)             n/a
  Total:            2                 4                      3
  
  Unreviewed:       1 ( 50%)          2 ( 50%)               3
  Needs-Work:       1 ( 50%)          2 ( 50%)               0
  
  Processing file : TOTAL (incomplete only):
  Type               Strings      Words (source)    Words (translation)
  Translated:       3 ( 75%)          8 ( 80%)               9
  Fuzzy:            1 ( 25%)          2 ( 20%)             n/a
  Untranslated:     0 (  0%)          0 (  0%)             n/a
  Total:            4                10                      9
  
  File count (incomplete):       1
  File count:       3
  
  
  '''
# ---
# name: test_output[no-color=True-incomplete=True-short-strings]
  '''
  tests/cli/data/test_pocount_po_fuzzy/one.po strings: total: 2	| 1t	1f	0u	| 50.0%t	50.0%f	0.0%u
  
  '''
# ---
# name: test_output[no-color=True-incomplete=True-short-words]
  '''
  tests/cli/data/test_pocount_po_fuzzy/one.po source words: total: 4	| 2t	2f	0u	| 50.0%t	50.0%f	0.0%u
  
  '''
# ---
